parser.rb

```diff
@@ -54,22 +55,23 @@ def parse_players
 
   
   def parse_kills
-    re_pattern_kill = /(?:Kill:.*:)(.*)(?:killed)/
+    re_pattern_kill = /(?:Kill:.*:)(.*)(?:killed)(.*)(?:by)/
     players = parse_players
     kills = players.each_with_object(Hash.new(0)){ |key, hash| hash[key] = 0}
     
     File.foreach(@filename_with_path) { |line|
       line.match(re_pattern_kill) { |m|
-        m.captures.each {|player|
-          kills[player.strip] += 1 if player
-        }
+        if m.captures[0] && m.captures[1]
+          if m.captures[0].strip == "<world>"
+            kills[m.captures[1].strip] -= 1
+          else
+            kills[m.captures[0].strip] += 1
+          end
+          @total_kills += 1
+        end         
```

Comment:
```rb
    all_kills = File.readlines(@filename_with_path).reduce({}) do |kills, line|
      killer, killed = line.match(re_pattern_kill).captures
      next unless killer && killed

      kills.merge(counting_kill(killer, killed))
    end
    
    @total_kills = all_kills.values.sum
```
e...
```rb
def counting_kill(killer, killed)
  { killed => kills[killed] - 1 } if killer == "<world>"
  { killer => kills[killer] + 1 }                  
end  
```

Reply:
Olá halan. Muito obrigado pela revisão. Eu testei a sua sugestão e acabei encontrando o seguinte problema ao executar o arquivo main.rb:  "undefined method `captures' for nil:NilClass (NoMethodError)". 

Esse comportamento ocorre quando a expressão regular não dá match com a linha do arquivo, ou seja, neste caso o match retorna nil. Consequentemente ele dá esse erro do ".captures " não existir na class "nil".

Reply:
Tem razão, o `match` é milindroso por retornar nil quando não dá match. Tem outra forma bacana.

```rb
next unless line =~ re_pattern_kill
kill, killer = $1, $2
```

Reply:
Referência: https://ruby-doc.org/core-3.0.2/Regexp.html#class-Regexp-label-Special+global+variables

Reply:
Um outro truque pra a a manga é esse:
https://ruby-doc.org/core-3.0.2/Regexp.html#class-Regexp-label-Special+global+variables

usar `named captures`, declara seu regex assim:

```rb
 re_pattern_kill = /(?:Kill:.*:) (?<killer>.*) (?:killed) (?<killed>.*) (?:by)/
```
utilizando parênteses com `(?<>)`

Então as capturas ficam assim:

```rb
next unless re_pattern_kill =~ line # é o suficiente pra declarar killer e killed
next unless killer && killed

```

Reply:
Mais truques com regex https://medium.com/rubycademy/3-ways-to-use-regexp-capture-groups-with-back-references-in-ruby-b4969cc9b3ec 🪄

Reply:
Resolvemos em pairing 🤝